models/Room.js

```diff
@@ -30,40 +30,51 @@ class Room {
      */
     constructor(roomIndex, roomMembers, maxPlayersPerRoom) {
         this.roomIndex = roomIndex
-        this.roomId =  uuidv4()
+        this.roomId = uuidv4()
         this.roomMembers = roomMembers
         this.maxPlayersPerRoom = maxPlayersPerRoom
         this.udpConnections = new Map()
     }
+
     /**
      * @param  {Player} player Player object, representing the player to be added to the room
      */
     addPlayer(player) {
         this.roomMembers.push(player)
     }
+
     /**
      * @param  {string} playerId unique player id assigned once the connection is established using uuid package
      */
     removePlayer(playerId) {
+        var updatedRoomMembers, counter
+        this.udpConnections.delete(playerId)
         this.roomMembers.forEach((roomMember, index, object) => {
             if (roomMember.playerId === playerId) {
                 object.splice(index, 1)
+                updatedRoomMembers = object
             }
         })
+        for (counter = 0; counter < updatedRoomMembers.length; counter++) {
+            updatedRoomMembers[counter].playerIndex = counter
+        }
+        this.roomMembers = updatedRoomMembers
```

Comment:
Is there a purpose for playerIndex? Why does it matter which order the room members entered the room?

If playerIndex isn't very important, I suggest getting rid of it and then changing roomMembers from an Array to a Set. This way look ups and deletion of a room member is O(1) instead of O(N).

Reply:
I think we talked about this before in terms of serialization and deserialization of Maps or Sets between the server and the client

Reply:
Could you remind me why it has to be an Array? And what playerIndex is used for?

Reply:
I will take FullRooms as an example.

Full rooms are array of Room Objects
If I used them as a Map, That is 3 levels of nested serialization
FullRoom Map -> Room -> Player

I can use the same Map to Dictionary through hashtable like Garner did with tags for C# SDK, But would your rather add this step on the client for every GetRooms or GetAvailableRooms(called often).

On the other hand, removing a room from fullRooms list is O(N) which is not perfect but the N here is usually a value between 1 and 10.
Default is 4 (MAX rooms per lobby)

Reply:
One use of PlayerIndex is if there any OrphanObservable
The default option is to let the 1st player in the room observe its movement and update other room members.

>OrphanObservable( A GameObject created in UnityEditor (offline) and we are interested to sync its position once the game starts ex.PingPong Ball)

Reply:
Why not a Hashset of FullRooms? Does that require extra steps before sending as well?

As for playerIndex, if you are just searching for the 1st player, it doesn't seem like it matters which player actually observes the object. So, you could just pick any player from the set and get rid of playerIndex entirely

Reply:
I think it's just because there's no such thing as a map util deserialized into one. You'd have to pack and unpack if it needs to ever be sent to/from the server, and JSON Parser already iterated through once to produce the array in the first place.